### PR TITLE
Refactor Task 1,2,3. Finished Task number 4,5,7 added 2 properties to…

### PR DESCRIPTION
**Pull request by zur250**
… the activity : min_time_to_start and max_time_to_start. Added 2 new methods to the project : calc activity min time and calc activity max time. finished calc activity min time.

### CODE REVIEW

File: Cpm.py [hunk 1]
```diff
@@ -4,43 +4,98 @@
...
+    def calculate_activity_min_time_to_start(self):
+        for activity, depending_activity in self._project.items():
+            for dep_activity in depending_activity:
+                if (dep_activity.min_time_to_start == 0 or
+                        dep_activity.min_time_to_start > activity.duration + activity.min_time_to_start):
+                    dep_activity.min_time_to_start = activity.duration + activity.min_time_to_start
+                    dep_activity.max_time_to_start = dep_activity.min_time_to_start
+
+    def calculate_activity_max_time_to_start(self):
+        '''for activity, depending_activity in self._project.items():
```
Comment (OrenZak) [inline Cpm.py]: why you have the ''' at the start and at the end ?

Comment (zur250) [reply Cpm.py]: it is a comment for multiple line. its not good yet

File: Cpm.py [hunk 2]
```diff
@@ -4,43 +4,98 @@
...
+        self.min_time_to_start = 0
+        self.max_time_to_start = 0
+        #self._depending_activities = []
 
     def __str__(self):
-        print("Activity %s with duration %s", self._name, self._duration)
+        print("Activity %s with duration %s", self.name, self.duration)
+
+    def __eq__(self, other):
+        return self.name == other.name and self.duration == other.duration
```
Comment (OrenZak) [inline Cpm.py]: not sure about it, maybe we should generate a UUID to each activity on create time and just compare it.

File: Cpm.py [hunk 3]
```diff
@@ -4,43 +4,98 @@
...
 
     def __init__(self, project):
+        # this method initializes a project object if no dictionary or None is given, an empty dictionary will be used
         if project is not None:
-            self._project = project
+            self.__project = project
+        else:
+            self.__project = {}
+
+        self.__project_duration = 0
```
Comment (OrenZak) [inline Cpm.py]: `__project_duration` should be calculated if there is a project content.

File: Cpm.py (hunk 2 above)
Comment (zur250) [reply Cpm.py]: it might be a good idea.

File: Cpm.py (hunk 3 above)
Comment (zur250) [reply Cpm.py]: i will midify it tomorrow

File: Cpm.py [hunk 4]
```diff
@@ -4,43 +4,98 @@
...
+
+    def add_depending_activity(self, activity, depending_activity):
+        # this method add depending activity to activity in the project
+        if activity in self.__project:
+            if depending_activity in self.__project[activity]:
+                self._project[activity].append(depending_activity)
+                logger.debug("%s added to the activity", depending_activity)
+            else:
+                logger.debug("%s already in the list of depending activities", depending_activity)
```
Comment (OrenZak) [inline Cpm.py]: maybe we should update, in case the activity content was changed

Comment (zur250) [reply Cpm.py]: it suppose to check the equals method that i implemented on the activity class.

File: Cpm.py [hunk 5]
```diff
@@ -4,43 +4,98 @@
...
-    def add_activity(self, name, duration):
-        self._project[name] = Activity(name, duration)
-        logger.debug("%s added to the project", self._project[name])
+    def add_activity(self, activity):
+        # this method add activity to the project
+        if activity not in self.__project:
+            self._project[activity] = []
+            logger.debug("%s added to the project", activity)
+        else:
+            logger.debug("%s already at the", activity)
```
Comment (OrenZak) [inline Cpm.py]: also here maybe we should update the activity

Comment (zur250) [reply Cpm.py]: it suppose to check the equals method that i implemented on the activity class.

File: Cpm.py [hunk 6]
```diff
@@ -4,43 +4,98 @@
...
 
     def remove_activity(self, activity):
-        if activity in self._project:
-            self._project.pop(activity)
+        # this method removing activity from the project
+        if activity in self.__project:
```
Comment (OrenZak) [inline Cpm.py]: lets log the `else` option: `logger.debug("%s trying to delete activity that not in the project", activity)`

Comment (zur250) [reply Cpm.py]: will modify tomorrow

File: Cpm.py (hunk 1 above)
Comment (OrenZak) [reply Cpm.py]: 👍

File: Cpm.py [hunk 7]
```diff
@@ -4,43 +4,98 @@
...
+        for activity, depending_activity in self._project.items():
+            for dep_activity in depending_activity:
+                if (dep_activity.min_time_to_start == 0 or
+                        dep_activity.min_time_to_start > activity.duration + activity.min_time_to_start):
+                    dep_activity.min_time_to_start = activity.duration + activity.min_time_to_start
+                    dep_activity.max_time_to_start = dep_activity.min_time_to_start
+
+    def calculate_activity_max_time_to_start(self):
+        '''for activity, depending_activity in self._project.items():
+            for dep_activity in depending_activity:
```
Comment (OrenZak) [inline Cpm.py]: rename second param to `depending_activities`

File: Cpm.py [hunk 8]
```diff
@@ -4,43 +4,98 @@
...
+                if (dep_activity.max_time_to_start == 0 or
+                        dep_activity.max_time_to_start < activity.duration + activity.min_time_to_start):
+                    dep_activity.min_time_to_start = activity.duration + activity.min_time_to_start'''
         pass
 
     def find_critical_path(self):
-        # todo find critical path of the project (Showing the edges of the critical pass with their length
-        pass
+        # find critical path of the project (Showing the edges of the critical pass with their length)
+        for activity in self.__project.keys():
```
Comment (OrenZak) [inline Cpm.py]: how do we make sure it starts form the first created activity?

**Review by OrenZak — APPROVED** [review]
see my comments 👍

File: Cpm.py (hunk 8 above)
Comment (zur250) [reply Cpm.py]: idk

File: Cpm.py (hunk 7 above)
Comment (zur250) [reply Cpm.py]: will modify

File: Cpm.py (hunk 3 above)
Comment (zur250) [reply Cpm.py]: done

File: Cpm.py (hunk 6 above)
Comment (zur250) [reply Cpm.py]: done

File: Cpm.py (hunk 7 above)
Comment (zur250) [reply Cpm.py]: done

**Review by OrenZak — APPROVED** [review]
👍

**Pull request merged**